### PR TITLE
<format>: Allow move-only output iterators

### PR DESCRIPTION
**Pull request by eldakesh-ms**
Now we move around the iterator we get instead of copying it. I had to
move a few functions from their home in algorithm and xmemory into
xutility. I hope I got the comments/guards correct, but please double
check those.

The testing is done by making sure we don't use-after-move a custom
iterator and that all the printing functions are instantiated.

### CODE REVIEW

**Comment by CaseyCarter** [comment]
> I had to move a few functions from their home in algorithm and xmemory into xutility.

Had to "move" them, eh? Good one!

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/format
Comment (miscco) [inline stl/inc/format:1179]: So we can only call out once?

File: stl/inc/xutility
Comment (miscco) [inline stl/inc/xutility:4552]: Can we take the opportunity t turn this into an early return?

File: stl/inc/xutility
Comment (miscco) [inline stl/inc/xutility:4554]: Do we have a convention whether the `if (!_STD is_constant_evaluated()) {` inside the `if constexpr` or the other way

I guess it avoids an empty if block when the optimizations are not availabe

File: stl/inc/xutility
Comment (miscco) [reply stl/inc/xutility:4239]: Didnt we have the memcpy optimization in place?

Comment (eldakesh-ms) [reply stl/inc/xutility:4239]: This was the only change with regards to functions being moved.

File: stl/inc/format
Comment (eldakesh-ms) [reply stl/inc/format:2024]: This happened because there were no explicit constraints in the standard for `format_to`. I wasn't sure how strong "equivalent to" meant.

File: stl/inc/format
Comment (eldakesh-ms) [reply stl/inc/format:1179]: Until we put set a new iterator, yes.

File: stl/inc/xutility
Comment (eldakesh-ms) [reply stl/inc/xutility:4552]: I don't feel too comfortable changing up any of the functions I moved (apart from the format change), I think most reviewers would assume I simply copy-pasted.

File: stl/inc/xutility
Comment (CaseyCarter) [reply stl/inc/xutility:4239]: > Didnt we have the memcpy optimization in place?

I think that's in #1772, which likely won't make 16.10p3, so we'll have to clean up the conflict after this merges.

File: stl/inc/format
Comment (CaseyCarter) [reply stl/inc/format:2024]: [[structure.specifications]/4](http://eel.is/c++draft/structure.specifications#4) defines precisely what "Equivalent to" means when it appears in an _Effects_ element:

Whenever the _Effects_ element specifies that the semantics of some function `F` are _Equivalent to_ some code sequence, then the various elements are interpreted as follows. If `F`'s semantics specifies any _Constraints_ or _Mandates_ elements, then those requirements are logically imposed prior to the equivalent-to semantics. Next, the semantics of the code sequence are determined by the _Constraints_, _Mandates_, _Preconditions_, _Effects_, _Synchronization_, _Postconditions_, _Returns_, _Throws_, _Complexity_, _Remarks_, and _Error conditions_ specified for the function invocations contained in the code sequence. The value returned from `F` is specified by `F`'s _Returns_ element, or if `F` has no _Returns_ element, a non-`void` return from `F` is specified by the return statements ([[stmt.return]](http://eel.is/c++draft/statement.return)) in the code sequence. If `F`'s semantics contains a _Throws_, _Postconditions_, or _Complexity_ element, then that supersedes any occurrences of that element in the code sequence.

File: stl/inc/xutility
Comment (CaseyCarter) [inline stl/inc/xutility:4160]: Woops - I must have been unclear. I didn't mean "move this line", I meant "move this entire code block that you've introduced". The order here should be:

1. `std::copy` overloads (and support machinery if any)
2.  `ranges::_Copy_unchecked`
3. `std::copy_n` overloads

with the idea being that if we move the rest of `ranges::copy` here where `ranges::_Copy_unchecked` is defined, all of this would be in the order declarations appear in the `<algorithm>` synopsis in the Standard.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/xutility
Comment (eldakesh-ms) [reply stl/inc/xutility:4160]: Right, that makes way more sense. I was a little confused as well.

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/format
Comment (CaseyCarter) [reply stl/inc/format:1179]: Hmm. This seems like a defect we need to fix in the Standard; ditto the _Effects_ of `advance_to`.

Comment (CaseyCarter) [reply stl/inc/format:1179]: I don't want to block merging this, so I've recorded an action item to submit an LWG issue.

**Comment by eldakesh-ms** [comment]
Squashed and rebased. Double checked that I didn't undo any `_NODISCARD`s but it's possible I did.

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp
Comment (barcharcraz) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:65]: let's use some other name for the template argument, or use capitals. Shadows a ::std:: function (thanks stl)

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp
Comment (barcharcraz) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:124]: we should test a std::formatter based type as well, although it doesn't have to be in this PR. (to prove it works you could add a `static_cast<void*>(nullptr)`)

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp
Comment (barcharcraz) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:64]: this may be simpler if it just always uses back_insert_iterator and is parameterized on it's type parameters (or just charT)

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp
Comment (barcharcraz) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:68]: Why do we need to have this member and do the asserts as we go, instead of just making the type move only and immediately aborting if we copy it?

Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:68]: oh, it tracks double moves, I see.

File: stl/inc/format
Comment (barcharcraz) [inline stl/inc/format:1944]: it might be safer if we rval qualify the op().

**Review by barcharcraz — CHANGES_REQUESTED** [review]

File: stl/inc/format
Comment (eldakesh-ms) [reply stl/inc/format]: Outdated.

**Comment by eldakesh-ms** [comment]
There is an outstanding issue with custom formatting. It's not a result of this PR, but running the tests without `/SDL` will cause issues.

**Comment by eldakesh-ms** [comment]
Rebased, double check the function moves as those were re-done (xutilty, xmemory, algorithm).

**Review by miscco — APPROVED** [review]
I am not 100% up to date with fmt but the changes look good for me

**Review by barcharcraz — APPROVED** [review]

**Pull request merged**
